client/src/app/queries/tags.ts

```diff
@@ -54,10 +54,12 @@ export const useCreateTagMutation = (
     onSuccess: (res) => {
       onSuccess(res);
       queryClient.invalidateQueries([TagsQueryKey]);
+      queryClient.invalidateQueries([TagCategoriesQueryKey]);
```

Comment:
Since most of the mutations seem to need to invalidate both queries, why not just have a helper to make sure things are handled that same way for all?

E.g.
```ts
const useInvalidatingMutation = <T>(
  mutationFn: MutationFunction<T, T>,
  onSuccess: (res: any) => void,
  onError: (err: AxiosError) => void
) => {
  const queryClient = useQueryClient();

  return useMutation({
    mutationFn,
    onSuccess: (res) => {
      onSuccess(res);
      queryClient.invalidateQueries([TagsQueryKey]);
      queryClient.invalidateQueries([TagCategoriesQueryKey])
    },
    onError: (err: AxiosError) => {
      onError(err);
      queryClient.invalidateQueries([TagsQueryKey]);
      queryClient.invalidateQueries([TagCategoriesQueryKey])
    },
  });
}
```

And then this mutation can just look like:
```ts
export const useCreateTagMutation = (
  onSuccess: (res: any) => void,
  onError: (err: AxiosError) => void
) => {
  return useInvalidatingMutation(createTag, onSuccess, onError);
};
```

Reply:
I would say we should perhaps include an additional `mutationOptions` object that is a `Partial<UseMutationOptions>` and gets spread into the inner useMutation, or perhaps just have the arguments for `useInvalidatingMutation` be themselves the same `UseMutationOptions` object, but the onSuccess and onError just get overridden? TypeScript generics might get weird I suppose.

Also, we may want to consider using compound query key arrays that contain some common element we can pass to invalidate any query that has to do with tags. The `queryKey` can be an array with any number of elements to identify the query, and if any of those elements is passed into `invalidateQueries` it will invalidate all matching queries. https://tanstack.com/query/v4/docs/react/guides/query-keys

That would at least mean we can do a single `queryClient.invalidateQueries([TagsQueryKey])` if we just include `TagsQueryKey` along with `TagCategoriesQueryKey` in the tag categories query hook. It still doesn't prevent all the duplication @sjd78's suggestion would address though, so maybe we can do both.

Reply:
Thanks @sjd78 and @mturley , for the review.

I've pushed an update using compound query key which simplifies and therefore it doesn't need to use an helper.
Effectively on the tags page, the tags are depending upon the tagCategories since we're displaying the tags for each tag category they belong to, therefore when a tag is touched (Created/updated/deleted) then we want to invalidate the same query key we use for fetching the tagCategories.